File: princeton-algorithms-part1/week1/dynamic-connectivity/unionfind/quickunion.py
import unionfind


class QuickUnion(unionfind.UnionFind):
    def __init__(self, N: int) -> None:
        self.id = list(range(N))

    def __root(self, i: int) -> int:
        while i != self.id[i]:
            i = self.id[i]
        return i

    def union(self, p: int, q: int) -> None:
        self.id[self.__root(p)] = self.__root(q)

    def connected(self, p: int, q: int) -> bool:
        return self.__root(p) == self.__root(q)


class WeightedQuickUnion(unionfind.UnionFind):
    def __init__(self, N: int) -> None:
        self.id = list(range(N))
        self.sz = [1] * N

    def __root(self, i: int) -> int:
        while i != self.id[i]:
            i = self.id[i]
        return i

    def union(self, p: int, q: int) -> None:
        pr = self.__root(p)
        qr = self.__root(q)

        if pr == qr:
            return

        if self.sz[pr] < self.sz[qr]:
            self.id[pr] = qr
            self.sz[qr] += self.sz[pr]
        else:
            self.id[qr] = pr
            self.sz[pr] += self.sz[qr]

    def connected(self, p: int, q: int) -> bool:
        return self.__root(p) == self.__root(q)


class WeightedQuickUnionPathCompression(unionfind.UnionFind):
    def __init__(self, N: int) -> None:
        self.id = list(range(N))
        self.sz = [1] * N

    def __root(self, i: int) -> int:
        while i != self.id[i]:
            self.id[i] = self.id[self.id[i]]
            i = self.id[i]
        return i

    def union(self, p: int, q: int) -> None:
        pr = self.__root(p)
        qr = self.__root(q)

        if pr == qr:
            return

        if self.sz[pr] < self.sz[qr]:
            self.id[pr] = qr
            self.sz[qr] += self.sz[pr]
        else:
            self.id[qr] = pr
            self.sz[pr] += self.sz[qr]

    def connected(self, p: int, q: int) -> bool:
        return self.__root(p) == self.__root(q)

File: princeton-algorithms-part1/week1/2-union-find/lab/QuickUnion.java
import java.util.Map;
import java.util.HashMap;


// QuickUnion implements the UnionFind interface
//
// Every time we union two elements we mark the representative element from one union as
// the parent of the representative element from the other union
//
// This makes the union operation fast as it only have to associate the two representative
// elements. However it will now depends on the find operation which can take up to O(n) time.

public class QuickUnion<T> implements UnionFind<T> {
  Map<T, T> parent = new HashMap<>();

  // return the representative element of the union containing the input element
  public T find(T e) {
    while (parent.containsKey(e))
      e = parent.get(e);

    return e;
  }

  // connect two elements combining their two unions into one
  public void union(T e1, T e2) {
    parent.put(find(e2), find(e1));
  }

  // return whether two elements are connected (belong to the same union)
  public boolean isConnected(T e1, T e2) {
    return find(e1).equals(find(e2));
  }

  public static void main(String[] args) {
    UnionFind<Character> uf = new QuickUnion<>();
    uf.union('a', 'b');
    uf.union('c', 'd');
    uf.union('e', 'f');
    uf.union('a', 'd');
    assert uf.isConnected('b', 'c') == true;
    assert uf.isConnected('f', 'c') == false;
  }
}
